cms/djangoapps/contentstore/views/assets.py

```diff
@@ -94,97 +109,168 @@ def _assets_json(request, course_key):
-    elif requested_sort == 'display_name':
-        requested_sort = 'displayname'
-    sort = [(requested_sort, sort_direction)]
+    filter_parameters = None
+
+    if (request_options['requested_asset_type']):
+        filter_parameters = _get_filter_parameters_for_mongo(request_options['requested_asset_type'])
+
+    sort_type_and_direction = _get_sort_type_and_direction(request_options)
 
-    current_page = max(requested_page, 0)
-    start = current_page * requested_page_size
-    options = {
+    requested_page_size = request_options['requested_page_size']
+    current_page = _get_current_page(request_options['requested_page'])
+    first_asset_to_display_index = _get_first_asset_index(current_page, requested_page_size)
+
+    query_options = {
         'current_page': current_page,
         'page_size': requested_page_size,
-        'sort': sort,
-        'filter_params': filter_params
+        'sort': sort_type_and_direction,
+        'filter_params': filter_parameters
     }
-    assets, total_count = _get_assets_for_page(request, course_key, options)
-    end = start + len(assets)
-
-    # If the query is beyond the final page, then re-query the final page so
-    # that at least one asset is returned
-    if requested_page > 0 and start >= total_count:
-        options['current_page'] = current_page = int(math.floor((total_count - 1) / requested_page_size))
-        start = current_page * requested_page_size
-        assets, total_count = _get_assets_for_page(request, course_key, options)
-        end = start + len(assets)
-
-    asset_json = []
-    for asset in assets:
-        asset_location = asset['asset_key']
-        # note, due to the schema change we may not have a 'thumbnail_location'
-        # in the result set
-        thumbnail_location = asset.get('thumbnail_location', None)
-        if thumbnail_location:
-            thumbnail_location = course_key.make_asset_key(
-                'thumbnail', thumbnail_location[4])
-
-        asset_locked = asset.get('locked', False)
-        asset_json.append(_get_asset_json(
-            asset['displayname'],
-            asset['contentType'],
-            asset['uploadDate'],
-            asset_location,
-            thumbnail_location,
-            asset_locked
-        ))
-
-    return JsonResponse({
-        'start': start,
-        'end': end,
+
+    assets, total_count = _get_assets_for_page(request, course_key, query_options)
+
+    if requested_page_size > 0 and first_asset_to_display_index >= total_count:
+        _update_options_to_requery_final_page(query_options, total_count)
+        first_asset_to_display_index = _get_first_asset_index(current_page, requested_page_size)
+        assets, total_count = _get_assets_for_page(request, course_key, query_options)
+
+
+    last_asset_to_display_index = first_asset_to_display_index + len(assets)
+    assets_in_json_format = _get_assets_in_json_format(assets, course_key)
+
+    response_payload = {
```

Comment:
I'd like to generate this payload via a  separate function, but I don't want to pass in all the necessary state. Any thoughts about the readability would be improved by extracting this dictionary initialization out into a function?

Reply:
Since there's no actual work (i.e. - the variables all contain the correct values already, they just need to be organized a bit) in the dictionary, I think it's fine to leave as-is. The added overhead of a separate function seems unnecessary IMO.